src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -1029,13 +1029,52 @@ public static IGraphQLBuilder AddValidationRule<TValidationRule>(this IGraphQLBu
         }
         #endregion
 
-        #region - AddMetrics -
+        #region - AddApolloTracing / AddMetrics -
+        /// <summary>
+        /// Registers <see cref="InstrumentFieldsMiddleware"/> within the dependency injection framework and
+        /// configures it to be installed within the schema, and configures responses to include Apollo
+        /// Tracing data when enabled via <see cref="ExecutionOptions.EnableMetrics"/>.
+        /// When <paramref name="enable"/> is <see langword="true"/>, configures execution to set
+        /// <see cref="ExecutionOptions.EnableMetrics"/> to <see langword="true"/>; otherwise leaves it unchanged.
+        /// </summary>
+        public static IGraphQLBuilder AddApolloTracing(this IGraphQLBuilder builder, bool enable = true)
+            => AddApolloTracing(builder, _ => enable);
+
+        /// <summary>
+        /// Registers <see cref="InstrumentFieldsMiddleware"/> within the dependency injection framework and
+        /// configures it to be installed within the schema, and configures responses to include Apollo
+        /// Tracing data when enabled via <see cref="ExecutionOptions.EnableMetrics"/>.
+        /// Configures execution to run <paramref name="enablePredicate"/> and when <see langword="true"/>, sets
+        /// <see cref="ExecutionOptions.EnableMetrics"/> to <see langword="true"/>; otherwise leaves it unchanged.
+        /// </summary>
+        public static IGraphQLBuilder AddApolloTracing(this IGraphQLBuilder builder, Func<ExecutionOptions, bool> enablePredicate)
+        {
+            if (enablePredicate == null)
+                throw new ArgumentNullException(nameof(enablePredicate));
+
+            builder.AddMiddleware<InstrumentFieldsMiddleware>();
+            builder.ConfigureExecution(async (options, next) =>
+            {
+                if (enablePredicate(options))
+                    options.EnableMetrics = true;
+                DateTime start = DateTime.UtcNow;
+                var ret = await next(options).ConfigureAwait(false);
+                if (options.EnableMetrics)
+                {
+                    ret.EnrichWithApolloTracing(start);
```

Comment:
I'm 70/30 to always add tracing key even if Metrics.Enabled is false and ExecutionResult.Perf is null if someone call AddApolloTracing.

Reply:
Maybe I'm wrong and if the `tracing` key exists then it should have some content.

Reply:
Typically a user wants to add tracing results only based on a configuration setting. I surely wouldn’t want my production systems adding trace results, unless I (for example) set a HTTP header for debugging. Pretty sure we do not want any added results if EnableMetrics is false.

Reply:
Right, OK.